u_http.go

```diff
@@ -0,0 +1,565 @@
+	return d.fn(context.Background(), network, address)
+}
+
+// DialContext is like net.Dialer.DialContext.
+func (d *uhttpForwardDialer) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	return d.fn(ctx, network, address)
+}
+
+// getDialContextFn returns the proper DialContext function to use. This
+// function will honor the configured ProxyURL, if any.
+func (txp *UHTTPTransport) getDialContextFn(ctx context.Context) (dialContextFn, error) {
+	dialFn := txp.DialContext
+	if dialFn == nil {
+		dialFn = (&net.Dialer{}).DialContext
+	}
+	proxyURL := uhttpContextWithProxyURL(ctx)
+	if proxyURL == nil {
+		return dialFn, nil
+	}
+	dialer, err := proxy.FromURL(proxyURL, &uhttpForwardDialer{dialFn})
+	if err != nil {
+		return nil, err
+	}
+	contextDialer, good := dialer.(proxy.ContextDialer)
+	if !good {
+		return nil, errors.New("uhttp: bug: cannot get a ContextDialer")
+	}
+	return contextDialer.DialContext, nil
+}
+
+// handshakeTimeout returns the TLS handshake timeout.
+func (txp *UHTTPTransport) tlsHandshakeTimeout() time.Duration {
+	if txp.TLSHandshakeTimeout > 0 {
+		return txp.TLSHandshakeTimeout
+	}
+	return 10 * time.Second
+}
+
+// utlsClientHelloID returns the utls.ClientHelloID
+// that we should be using for the handshake.
+func (txp *UHTTPTransport) utlsClientHelloID() ClientHelloID {
+	if txp.UTLSClientHelloID != nil {
+		return *txp.UTLSClientHelloID
+	}
+	return HelloFirefox_Auto
+}
+
+// connCachePop extracts one of the connections in the cache
+// that are indexed by the provided address. Returns nil if
+// we don't have any entry in cache for the address.
+func (txp *UHTTPTransport) connCachePop(address string) net.Conn {
+	defer txp.mu.Unlock()
+	txp.mu.Lock()
+	if cl, found := txp.connCache[address]; found && len(cl) >= 1 {
+		conn := cl[0]
+		cl = cl[1:]
+		if len(cl) >= 1 {
+			txp.connCache[address] = cl
+		} else {
+			delete(txp.connCache, address) // don't keep empty cache entries
+		}
+		uhttpLog.Printf("uhttp: connCache pop %s => conn#%s", address, conn.RemoteAddr())
+		return conn
+	}
+	return nil
+}
+
+// CloseIdleConnections allows an http.Client controlling this
+// transport to close the idle connections.
+func (txp *UHTTPTransport) CloseIdleConnections() {
+	// Implementation note: cached connections are also
+	// cleaned up. Consider the case of a request that is
+	// interrupted via the context after it caused us to
+	// create a cached conn and before the RoundTripper has
+	// a chance to use the conn. Consider that after that
+	// the user calls CloseIdleConnections and then the
+	// transport goes out of scope. In such a case,
+	// we clearly want to get rid of the cached conn,
+	// otherwise we would leak the open conns.
+	if txp.cleartext != nil {
+		txp.cleartext.CloseIdleConnections()
+	}
+	if txp.https != nil {
+		txp.https.CloseIdleConnections()
+	}
+	if txp.h2 != nil {
+		txp.h2.CloseIdleConnections()
+	}
+	defer txp.mu.Unlock()
+	txp.mu.Lock()
```

Comment:
Don't we need to lock earlier?

Reply:
If goroutine A calls `RoundTrip` and goroutine B calls `CloseIdleConnections` concurrently, we have a data race because the write on, say, `txp.h2` is not guaranteed to happen before the read that we do here.

(I would still like the `CloseIdleConnections` to go unlocked, because it does not need to synchronize with any other synchronized access performed by this structure, but I should refactor the code to avoid this race.)